keps/sig-node/4576-conditions-for-pod-terminated/README.md

```diff
@@ -0,0 +1,906 @@
+
+To get started with this template:
+
+- [ ] **Pick a hosting SIG.**
+  Make sure that the problem space is something the SIG is interested in taking
+  up. KEPs should not be checked in without a sponsoring SIG.
+- [ ] **Create an issue in kubernetes/enhancements**
+  When filing an enhancement tracking issue, please make sure to complete all
+  fields in that template. One of the fields asks for a link to the KEP. You
+  can leave that blank until this KEP is filed, and then go back to the
+  enhancement and add the link.
+- [ ] **Make a copy of this template directory.**
+  Copy this template into the owning SIG's directory and name it
+  `NNNN-short-descriptive-title`, where `NNNN` is the issue number (with no
+  leading-zero padding) assigned to your enhancement above.
+- [ ] **Fill out as much of the kep.yaml file as you can.**
+  At minimum, you should fill in the "Title", "Authors", "Owning-sig",
+  "Status", and date-related fields.
+- [ ] **Fill out this file as best you can.**
+  At minimum, you should fill in the "Summary" and "Motivation" sections.
+  These should be easy if you've preflighted the idea of the KEP with the
+  appropriate SIG(s).
+- [ ] **Create a PR for this KEP.**
+  Assign it to people in the SIG who are sponsoring this process.
+- [ ] **Merge early and iterate.**
+  Avoid getting hung up on specific details and instead aim to get the goals of
+  the KEP clarified and merged quickly. The best way to do this is to just
+  start with the high-level sections and fill out details incrementally in
+  subsequent PRs.
+
+Just because a KEP is merged does not mean it is complete or approved. Any KEP
+marked as `provisional` is a working document and subject to change. You can
+denote sections that are under active debate as follows:
+
+```
+<<[UNRESOLVED optional short context or usernames ]>>
+Stuff that is being argued.
+<<[/UNRESOLVED]>>
+```
+
+When editing KEPS, aim for tightly-scoped, single-topic PRs to keep discussions
+focused. If you disagree with what is already in a document, open a new PR
+with suggested changes.
+
+One KEP corresponds to one "feature" or "enhancement" for its whole lifecycle.
+You do not need a new KEP to move from beta to GA, for example. If
+new details emerge that belong in the KEP, edit the KEP. Once a feature has become
+"implemented", major changes should get new KEPs.
+
+The canonical place for the latest set of instructions (and the likely source
+of this file) is [here](/keps/NNNN-kep-template/README.md).
+
+**Note:** Any PRs to move a KEP to `implementable`, or significant changes once
+it is marked `implementable`, must be approved by each of the KEP approvers.
+If none of those approvers are still appropriate, then changes to that list
+should be approved by the remaining approvers and/or the owning SIG (or
+SIG Architecture for cross-cutting KEPs).
+-->
+# KEP-4576: Condition PodTerminated added for pod
```

Comment:
Main concern with this proposal so far is the name of this condition.

Terminated in this case seems to be mean that the pod is either succeeded or failed. If a pod is preempted/disrupted/manually deleted, then I also consider that terminating. I think we may move these to failed (@mimowo ?) but I find it weird that this condition would be "PodTerminated" if it is Failed Or Succeeded.

Generally I think you want this condition because you want to say that the pod resources are ready to be cleaned up in case of pvc manager.

Names for these are always tricky so I may suggest a few options in the review.

I propose `PodComplete` or `PodTerminal` as an option.

Reply:
Mostly that are all cases where pods are terminated (deleted, preempted) mean that we can safetly assume that they are represented by Failed or Succeeded state.

Reply:
> > Terminated in this case seems to be mean that the pod is either succeeded or failed.
> > ...
> > I propose PodComplete or PodTerminal as an option.
> 
> Personally I like `PodTerminated` as it is concise, and IIUC the resource cleanups happen in [`SyncTerminatedPod`](https://github.com/kubernetes/kubernetes/blob/cae35dba5a3060711a2a3f958537003bc74a59c0/pkg/kubelet/kubelet.go#L2137). Alternative concise name that comes to my mind is `PodFinished`, which would also correspond to the kubelet's internal naming [here](https://github.com/kubernetes/kubernetes/blob/cae35dba5a3060711a2a3f958537003bc74a59c0/pkg/kubelet/status/status_manager.go#L63-L64).

+1

Reply:
There is an additional need to convey when a pod starts terminating in order to remove it from endpoints.  In that [problem description](https://docs.google.com/document/d/1t25jgO_-LRHhjRXf4KJ5xY_t8BZYdapv7MDAxVGY6R8/edit?usp=sharing) I propose an approach that would support what this KEP requests as well as the issue that all terminating pods need to be removed from endpoints to prevent traffic from being sent to already terminated pods.